Add tooling to automate release-based job forking

Katharine opened the pull request:
Release job forking is currently a PITA and it really doesn't have to be. Also I couldn't sleep.

Assumptions:

- Jobs for a release branch don't change (much) once they're set
- Jobs for a release branch may differ in arbitrary ways from the master version of the job, and eachother

Solution:

- Use regexes for `skip_branches` instead of enumerating them
- Centralise all the forked jobs together in one or two files per release
- Add an annotation for jobs that need to be forked
- Write a tool to automatically generate the job config files for a given release, making appropriate substitutions
- Generate required testgrid config changes

Thus far, this is implemented for presubmits, but not postsubmits (because some of those are pretty complicated). In order to actually do this, I made the following concrete changes (so far):

- In #12463: added `omitempty` to some Presubmit fields that are generally omitted and empty (`Agent`, `AlwaysRun`, `Trigger`, `RerunCommand`, `Context`)
- Added an `Annotations` field to prowjobs for storing arbitrary key-value pairs that prow itself doesn't care about
- In #12462: fixed the config validation of `skip_branches` settings that contain regexes
- Refactored some of the prow config loading code so I could pull out a `ReadJobConfig` public function that would give me unmangled config data without me having to write my own loader
- Wrote a tool that can generate forked versions of presubmits, postsubmits, and periodics.

Still to do:

- Generate testgrid config for those CI jobs
  - I intend to achieve this by making the testgrid configurator read some information out of prow annotations
- Split out the config and actually use the tool

/assign @krzyzacy @fejta @stevekuznetsov @BenTheElder @spiffxp @cjwagner

cjwagner commented:
/cc @imkin

stevekuznetsov commented on experiment/config-forker/main.go:
```diff
@@ -0,0 +1,76 @@
+	newPresubmits := map[string][]config.Presubmit{}
+	for repo, presubmits := range allPresubmits {
+		for _, presubmit := range presubmits {
+			if presubmit.Annotations["fork-per-release"] == "true" {
```
nit: should be constants

Katharine replied:
fair, will fix up when this entire thing gets refactored to death in the process of making postsubmits work.

stevekuznetsov commented on prow/config/config.go:
```diff
@@ -434,52 +434,27 @@ func Load(prowConfig, jobConfig string) (c *Config, err error) {
-	}
-
-	stat, err = os.Stat(jobConfig)
+// ReadJobConfig reads the JobConfig yaml, but does not expand or validate it.
```
You want this since you don't care to start the config agent loader? Seems like you could have gotten away with just starting it and reading once, but SGTM

Katharine replied:
If the config agent will give me the config without expansions (I didn't check) I am probably happy to use that instead — my issue is that I'm using what I read in to write it back out later. If expansion happens, then the result is a) hard to read, and also b) actually fails validation (because the presets are still attached and cause duplication errors when re-expanded).

I believe you have used this tooling to parse/write back your files, which suggests I've missed something.

BenTheElder replied:
It won't, there's a bunch of computed stuff from the config agent loading. The generated security jobs had a lot of problems with this

stevekuznetsov replied:
We don't use presets, so we side-stepped the issue.

stevekuznetsov commented on prow/config/jobs.go:
```diff
@@ -85,7 +85,7 @@ type JobBase struct {
 	MaxConcurrency int `json:"max_concurrency,omitempty"`
 	// Agent that will take care of running this job.
-	Agent string `json:"agent"`
+	Agent string `json:"agent,omitempty"`
```
These are great -- OpenShift had caught most but we never stopped setting agent when Cole defaulted it!

stevekuznetsov commented on prow/config/jobs.go:
```diff
@@ -113,7 +115,7 @@ type Presubmit struct {
 
 	// AlwaysRun automatically for every PR, or only when a comment triggers it.
-	AlwaysRun bool `json:"always_run"`
+	AlwaysRun bool `json:"always_run,omitempty"`
```
Can we doc the default if unset?

stevekuznetsov commented on prow/config/jobs.go:
```diff
@@ -209,7 +211,7 @@ type RegexpChangeMatcher struct {
 type Reporter struct {
 	// Context is the name of the GitHub status context for the job.
-	Context string `json:"context"`
+	Context string `json:"context,omitempty"`
```
Can we doc the default if unset?

BenTheElder commented on config/jobs/kubernetes/sig-aws/kops/kops-presubmits.yaml:
```diff
@@ -1,15 +1,13 @@
-    - release-1.13  # per-release image
-    - release-1.12  # per-release image
-    - release-1.11  # per-release image
+    - release-\d+\.\d+  # per-release image
```
👍

krzyzacy commented:
One question: what if the test config is different across different releases, or a certain featured presubmit job only works for new releases? How do we overwrite that?

Katharine commented:
@krzyzacy The idea is that you run the tool once, to generate a new config file for a release branch — after that, they're never touched again. The result is that the `release-1.15.yaml` or similar file is a tweaked copy of what was on master, and then never changed again.

This means that new tests, and changes to old tests, don't get added to old branches. Similarly, tests that are removed still run on old branches. I think this is correct behaviour.

If humans want to modify tests on old branches for whatever reason, they still have that option, since automation doesn't touch the files again.

krzyzacy commented:
that sounds :+1: ! So the commit https://github.com/kubernetes/test-infra/pull/12443/commits/4d1579addb3cf4c5dedc56856e2d90deba5b076f was done via copy-pasta and not by the generator?

Katharine commented:
Despite the commit message, yes (I should probably reword that).

krzyzacy commented:
Another qq: Currently each sig manages their own release-branch jobs (via the sig-foo subdir), how do we grant ownership via a single config file? Or we just say people should not touch these jobs in general and sig-release will own all release-branch jobs?

Katharine commented:
My thoughts are that people shouldn't need to touch these files (tests on old branches shouldn't change, I think, almost by definition?), and hopefully the frequency with which they do so is sufficiently low that us needing to approve the changes is not a problem (and perhaps it will even prevent changes that shouldn't be happening…).

krzyzacy commented:
Also looks like the -integration and -verify ones were missing (I meant to flip the integration one to the -podutil variant before last code freeze... I probably should do that soon)

Katharine commented:
I haven't done periodics or postsubmits yet (I'm working on them now) — currently just the presubmits are up.

(If there are presubmits I missed, oops! Can you point them out? I used the `manual-release-bump-required` comment as my tag for finding them)

krzyzacy commented:
https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-testing/integration.yaml and https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-testing/verify.yaml, they are... a bit different, but I think it's ok to generate them from 1.14 onwards

fejta commented on experiment/config-forker/BUILD.bazel:
```diff
@@ -0,0 +1,32 @@
+)
+
+go_binary(
+    name = "config-forker",
```
Can you make this more incomprehensible, nautical and/or greek?

Katharine replied:
`hippopotamus` it is.

fejta commented on experiment/config-forker/main.go:
```diff
@@ -0,0 +1,76 @@
+
+func parseFlags() options {
+	o := options{}
+	flag.StringVar(&o.jobConfig, "job-config", "config/", "Path to the job config")
```
Please don't provide a default value for --job-config

fejta commented on experiment/config-forker/main.go:
```diff
@@ -0,0 +1,76 @@
+	o := options{}
+	flag.StringVar(&o.jobConfig, "job-config", "config/", "Path to the job config")
+	flag.StringVar(&o.outputPath, "output", "", "Path to the output yaml")
+	flag.StringVar(&o.newVersion, "version", "", "Version number to generate jobs for")
```
blow up if this isn't a good value?

Katharine replied:
will clean up this whole file after I refactor for postsubmits.

fejta commented on experiment/config-forker/main.go:
```diff
@@ -0,0 +1,76 @@
+	}
+	allPresubmits := c.Presubmits
+	newPresubmits := map[string][]config.Presubmit{}
+	for repo, presubmits := range allPresubmits {
```
Don't we also have forked postsubmits and periodics?

Katharine replied:
Yup, supporting those is still WIP.

fejta commented:
What do you think about just submitting the go code for now and running it to convert the jobs later?

Katharine commented:
The job conversion is supposed to be a one-off for each version (for reasons discussed with @krzyzacy), so I can't just run the generator because what's on master does not match what these jobs actually do (the changes in this PR were generated by cut/paste).

I'm happy to submit the config changes separately, though.

Katharine commented:
I've added support for most periodics and postsubmits. There are still two periodics left behind:

* `ci-kubernetes-gce-conformance-stable-*` — because it lacks a root job to be generated from
* `ci-kubernetes-e2e-kubeadm-gce-upgrade-*-*` — because it has a bunch of complex version number use

There are now three annotations:

* `fork-per-release`: `true` will include the job in the forking
* `fork-per-release-periodic-interval`: if specified, an interval that will override the current one when forking.
* `fork-per-release-extra-arg-replacements`: if specified, a freeform set of replacements in the arguments e.g. `"latest-bazel.txt -> latest-bazel-{{.Version}}.txt"`

Additionally, some jobs got renamed:

* `ci-kubernetes-gce-conformance` -> `ci-kubernetes-gce-conformance-latest` (for consistency with the `ci-kubernetes-gce-conformance-*` tests it forks to)
* `ci-kubernetes-node-kubelet-{beta,stable1,stable2,stable3}` -> `ci-kubernetes-node-kubelet-{1.14,1.13,1.12,1.11}`
* `ci-kubernetes-node-kubelet-beta-features` -> `ci-kubernetes-node-kubelet-features-1.14`: reordered for consistency plus beta->1.14
* `ci-kubernetes-integration-{beta,stable1,stable2,stable3}` -> `ci-kubernetes-integration-{1.14,1.13,1.12,1.11}`
* `ci-kubernetes-verify-{beta,stable1,stable2,stable3}` -> `ci-kubernetes-verify-{1.14,1.13,1.12,1.11}`
* `ci-kubernetes-build-{beta,stable1,stable2,stable3}` -> `ci-kubernetes-build-all-{1.14,1.13,1.12,1.11}`: added `-all` to avoid conflicting with `ci-kubernetes-build`, which is completely different

Also, `ci-kubernetes-build-all` is new, to act as the template for `ci-kubernetes-build-all-{1.14,1.13,1.12,1.11}` (which had no equivalent on master).

I'm sure some of these changes merit some discussion somewhere.

I haven't yet changed these on testgrid (thus the test failure), or produced automation to generate updated testgrid config.

Katharine commented:
Having thought further about this, I think I'd like to handle Testgrid configuration separately.

To that end, I'd like to proceed with this like so:

1. Get agreement in principle from the necessary parties that a) the above renames are acceptable, and b) sig-release owning all the old-branch jobs (at least from an OWNERS perspective) is reasonable.
2. Revert the config moves, but leave the newly-added annotations in place (which are no-ops as far as prow is concerned)
3. Get this code merged
4. Design/implement a testgrid config solution, most likely utilising the annotation support added here (I have some ideas)

Aside from sig-release, I'm not sure who the necessary parties in 1) are.

Katharine commented:
I think this is now ready for review.

stevekuznetsov commented:
Annotations and ReadJobConfig 

/lgtm

stevekuznetsov commented:
/hold

spiffxp commented:
>Get agreement in principle from the necessary parties that a) the above renames are acceptable, and b) sig-release owning all the old-branch jobs (at least from an OWNERS perspective) is reasonable.

/lgtm
I am onboard with this FWIW

job renames = "loss" of history as far as some of our tooling goes (testgrid, triage, any metrics generated from bigquery), but I'm not sure how much this matters given that we've been rotating the meaning of buckets every 90 days (and logs age out after 90 days)

sig-release owning set of jobs as defined at branch time != sig-release being responsible to fix jobs when broken, this seems like an acceptable compromise

listx commented on experiment/config-forker/main.go:
```diff
@@ -0,0 +1,317 @@
+	if o.newVersion == "" {
+		return errors.New("--version must be specified")
+	}
+	if match, err := regexp.MatchString(`^\d+.\d+$`, o.newVersion); err != nil || !match {
```
I think you meant `\.` and not `.` (wildcard)

Katharine replied:
Fixed.

imkin commented:
/lgtm

stevekuznetsov commented on experiment/config-forker/README.md:
```diff
@@ -0,0 +1,64 @@
+
+## Usage
+
+* `--job-config`: Path to the job config
```
Does it need `--config-path` and `--job-config`?

Katharine replied:
No, it only uses the job config.

stevekuznetsov commented on experiment/config-forker/README.md:
```diff
@@ -0,0 +1,64 @@
+
+## Actions taken
+
+For the sake of clarity, all the below will assume we are forking for release-1.15.
```
Maybe clearer to say `--version 1.15`

Katharine replied:
Done.

spiffxp commented:
@stevekuznetsov what is necessary to remove the /hold here?

spiffxp commented:
/lgtm
/hold cancel
@stevekuznetsov please ping me if I should not have removed this

spiffxp commented:
ref: https://github.com/kubernetes/test-infra/issues/11977
